### PR TITLE
[Communication]: SMS 1:N Messages, Custom Tags, and Idempotence

### PR DESCRIPTION
**Pull request by lsundaralingam**
API View: https://apiview.dev/Assemblies/Review/f5d7cd8246734532ac59a09de84660a9

### CODE REVIEW

File: sdk/communication/azure-communication-sms/README.md [hunk 1]
```diff
@@ -54,23 +54,27 @@ endpoint = os.getenv('AZURE_COMMUNICATION_SERVICE_ENDPOINT')
...
+from azure.communication.sms import SendSmsOptions
 
-smsresponse = sms_client.send(
+sms_responses = sms_client.send(
     from_phone_number=PhoneNumberIdentifier("<leased-phone-number>"),
-    to_phone_number=[PhoneNumberIdentifier("<to-phone-number>")],
+    to_phone_numbers=["<to-phone-number-1>", "<to-phone-number-2>", "<to-phone-number-3>"],
     message="Hello World via SMS",
-    send_sms_options=SendSmsOptions(enable_delivery_report=True)) # optional property
+    enable_delivery_report=True, # optional property
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-sms/README.md:68]: I guess I am confused now, enableDeliveryReport is not optional anymore, right?

File: sdk/communication/azure-communication-sms/azure/communication/sms/_generated/_configuration.py [hunk 2]
```diff
@@ -38,7 +38,7 @@ def __init__(
         super(AzureCommunicationSMSServiceConfiguration, self).__init__(**kwargs)
 
         self.endpoint = endpoint
-        self.api_version = "2020-07-20-preview1"
+        self.api_version = "2021-03-07"
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-sms/azure/communication/sms/_generated/_configuration.py:41]: We should double check if we are going to remove preview now or later

File: sdk/communication/azure-communication-sms/azure/communication/sms/_generated/aio/_configuration.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def __init__(
         super(AzureCommunicationSMSServiceConfiguration, self).__init__(**kwargs)
 
         self.endpoint = endpoint
-        self.api_version = "2020-07-20-preview1"
+        self.api_version = "2021-03-07"
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-sms/azure/communication/sms/_generated/aio/_configuration.py:36]: Same comment as above

File: sdk/communication/azure-communication-sms/azure/communication/sms/_generated/aio/operations/_sms_operations.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 from azure.core.pipeline import PipelineResponse
 from azure.core.pipeline.transport import AsyncHttpResponse, HttpRequest
 
-from ... import models
+from ... import models as _models
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-sms/azure/communication/sms/_generated/aio/operations/_sms_operations.py:15]: Not following why this changed?

File: sdk/communication/azure-communication-sms/azure/communication/sms/_sms_client.py [hunk 5]
```diff
@@ -72,31 +78,55 @@ def from_connection_string(cls, conn_str,  # type: str
...
 
         request = SendMessageRequest(
-            from_property=from_phone_number.phone_number,
-            to=[p.phone_number for p in to_phone_numbers],
+            from_property=from_,
+            sms_recipients=[
+                SmsRecipient(
+                    to=p,
+                    repeatability_request_id=str(uuid4()),
+                    repeatability_first_sent=datetime.now()
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-sms/azure/communication/sms/_sms_client.py:114]: Anything needed to make it cultural neutral or is that not a problem?

File: sdk/communication/azure-communication-sms/samples/sms_sample.py [hunk 6]
```diff
@@ -31,13 +30,30 @@ def send_sms(self):
...
+            enable_delivery_report=True, # optional property
+            tag="custom-tag") # optional property
+        
+        failed_recipients = []
+        for sms_response in sms_responses:
+            if (sms_response.successful):
+                print("Message with message id {} was successful sent to {}"
+                .format(sms_response.message_id , sms_response.to))
+            else:
+                print("Message failed to send to {}"
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-sms/samples/sms_sample.py:46]: Will be good to print the actual code and errorMessage

File: sdk/communication/azure-communication-sms/samples/sms_sample.py [hunk 7]
```diff
@@ -31,13 +30,30 @@ def send_sms(self):
...
+                .format(sms_response.to))
+                failed_recipients.append(sms_response.to)
+        
+        # calling send() with failed recipients
+        sms_responses = sms_client.send(
+            from_="<leased-phone-number>",
+            to=failed_recipients,
+            message="Hello World via SMS",
+            enable_delivery_report=True, # optional property
+            tag="custom-tag") # optional property
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-sms/samples/sms_sample.py:56]: I know its a bit of overkill, but maybe we should not retry for bad requests, 400?

File: sdk/communication/azure-communication-sms/samples/sms_sample.py (hunk 6 above)
Comment (lsundaralingam) [reply sdk/communication/azure-communication-sms/samples/sms_sample.py:46]: Good idea, I'll update this.

File: sdk/communication/azure-communication-sms/samples/sms_sample.py (hunk 7 above)
Comment (lsundaralingam) [reply sdk/communication/azure-communication-sms/samples/sms_sample.py:56]: Sure, I can filter out the 400 requests before resending.

File: sdk/communication/azure-communication-sms/azure/communication/sms/_sms_client.py (hunk 5 above)
Comment (lsundaralingam) [reply sdk/communication/azure-communication-sms/azure/communication/sms/_sms_client.py:114]: Yes, this should be ```datetime.utcnow()``` , so it uses a standard time zone. We're also using UTC in other languages.

File: sdk/communication/azure-communication-sms/azure/communication/sms/_generated/aio/operations/_sms_operations.py (hunk 4 above)
Comment (lsundaralingam) [reply sdk/communication/azure-communication-sms/azure/communication/sms/_generated/aio/operations/_sms_operations.py:15]: I think its part of the newer autorest version. The [identity](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/communication/azure-communication-identity/azure/communication/identity/_generated/operations/_communication_identity_operations.py#L15) and [phone number](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/_generated/operations/_phone_number_administration_operations.py#L16) packages in python also do it this way.

File: sdk/communication/azure-communication-sms/azure/communication/sms/_generated/_configuration.py (hunk 2 above)
Comment (lsundaralingam) [reply sdk/communication/azure-communication-sms/azure/communication/sms/_generated/_configuration.py:41]: We discussed this offline and decided to leave it alone.

File: sdk/communication/azure-communication-sms/azure/communication/sms/_generated/aio/_configuration.py (hunk 3 above)
Comment (lsundaralingam) [reply sdk/communication/azure-communication-sms/azure/communication/sms/_generated/aio/_configuration.py:36]: We discussed this offline and decided to leave it alone.

File: sdk/communication/azure-communication-sms/README.md (hunk 1 above)
Comment (lsundaralingam) [reply sdk/communication/azure-communication-sms/README.md:68]: It's optional to the client, but we'll default to ```false``` if its not provided.

File: sdk/communication/azure-communication-sms/azure/communication/sms/_sms_client.py [hunk 8]
```diff
@@ -72,31 +78,58 @@ def from_connection_string(cls, conn_str,  # type: str
...
         :param str message: The message in the SMS
-        :keyword send_sms_options: the options object to configure delivery reporting.
-        :type send_sms_options: ~azure.communication.sms.models.SendSmsOptions
-        :return: The response object with the message_id
-        :rtype: SendMessageResponse: ~azure.communication.sms.models.SendMessageResponse
+        :keyword bool enable_delivery_report: Enable this flag to receive a delivery report for this
+         message on the Azure Resource EventGrid.
+        :keyword str tag: Use this field to provide metadata that will then be sent back in the corresponding
+         Delivery Report.
+        :return: A list of SmsSendResults.
```
Comment (ankitarorabit) [inline sdk/communication/azure-communication-sms/azure/communication/sms/_sms_client.py:96]: nit: SmsSendResult

**Review by ankitarorabit — APPROVED** [review]

**Pull request merged**